tests/composite_fk/models/tenant.py

```diff
@@ -0,0 +1,24 @@
+from django.db import models
+
+
+class Tenant(models.Model):
+    pass
+
+
+class User(models.Model):
+    tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ("tenant_id", "id")
+
+
+class Comment(models.Model):
+    tenant = models.ForeignKey(Tenant, on_delete=models.CASCADE)
+    user_id = models.IntegerField()
+    user = models.ForeignKey(
```

Comment:
I tried adding these models to a sandbox project, running `makemigrations` and then adding and running this test:

```python
from django.test import TestCase

from .models import Tenant, User, Comment


class TestFK(TestCase):
    def test_fk(self):
        tenant = Tenant.objects.create()
        user = User.objects.create(tenant=tenant)

        comment = Comment.objects.create(user=user)

        self.assertEqual(comment.user, user)
```

This resulted in this exception:

```python
python manage.py test fk
Found 1 test(s).
Creating test database for alias 'default'...
Traceback (most recent call last):
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 103, in _execute
    return self.cursor.execute(sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 350, in execute
    return super().execute(query)
           ^^^^^^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: duplicate column name: user_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lily/dev/sandbox/manage.py", line 22, in <module>
    main()
  File "/home/lily/dev/sandbox/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/commands/test.py", line 24, in run_from_argv
    super().run_from_argv(argv)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/commands/test.py", line 63, in handle
    failures = test_runner.run_tests(test_labels)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/test/runner.py", line 1066, in run_tests
    old_config = self.setup_databases(
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/test/runner.py", line 964, in setup_databases
    return _setup_databases(
           ^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/test/utils.py", line 206, in setup_databases
    connection.creation.create_test_db(
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/base/creation.py", line 78, in create_test_db
    call_command(
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/__init__.py", line 194, in call_command
    return command.execute(*args, **defaults)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/base.py", line 107, in wrapper
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/core/management/commands/migrate.py", line 357, in handle
    post_migrate_state = executor.migrate(
                         ^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
            ^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/migrations/executor.py", line 255, in apply_migration
    state = migration.apply(state, schema_editor)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/migrations/operations/models.py", line 97, in database_forwards
    schema_editor.create_model(model)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/base/schema.py", line 503, in create_model
    self.execute(sql, params or None)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/base/schema.py", line 202, in execute
    cursor.execute(sql, params)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 103, in _execute
    return self.cursor.execute(sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 350, in execute
    return super().execute(query)
           ^^^^^^^^^^^^^^^^^^^^^^
django.db.utils.OperationalError: duplicate column name: user_id
```

I was able to work around this by renaming `user_id` (to `user_id_inner`) and re-creating migrations from scratch, but then got this exception:

```python
python manage.py test fk
Found 1 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
E
======================================================================
ERROR: test_fk (fk.tests.TestFK.test_fk)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 354, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.IntegrityError: NOT NULL constraint failed: fk_comment.user_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lily/dev/sandbox/fk/tests.py", line 11, in test_fk
    comment = Comment.objects.create(user=user)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/query.py", line 679, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/base.py", line 863, in save
    self.save_base(
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/base.py", line 979, in save_base
    updated = self._save_table(
              ^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/base.py", line 1142, in _save_table
    results = self._do_insert(
              ^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/base.py", line 1183, in _do_insert
    return manager._insert(
           ^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/query.py", line 1847, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1831, in execute_sql
    cursor.execute(sql, params)
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lily/.local/share/lilyenv/virtualenvs/sandbox/3.12/lib/python3.12/site-packages/django/db/backends/sqlite3/base.py", line 354, in execute
    return super().execute(query, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django.db.utils.IntegrityError: NOT NULL constraint failed: fk_comment.user_id

----------------------------------------------------------------------
Ran 1 test in 0.002s

FAILED (errors=1)
Destroying test database for alias 'default'...
```

Reply:
Did you use this branch when testing? I'll look into it later.

Reply:
I did

Reply:
It was because the `from_fields`, `to_fields` were not serialized in the migrations. I pushed the fix for this. Thanks for reporting!